# JS2

## Promises

En *ES6* se puede manejar la asincronía utilizando el objeto **Promise**

Las promesas son **objetos** que representan el resultado final de una operación *asincrónica*. Un objeto ```Promise``` puede estar en uno de tres estados:

- *Pendiente* (**pending**): estado inicial mientras el valor aún no está disponible, ni cumplida ni rechazada.
- *Cumplida* (**fulfilled**): significa que la operación se completó satisfactoriamente. Por ejemplo, la promesa de una solicitud podría resolverse con un objeto JSON como su valor.
- *Rechazada* (**rejected**): La operación ha fallado y la promesa tiene una razón para el fracaso. Esta razón suele ser un ```error``` de algún tipo

El estado inicial de una promesa asincrónica es *pending*, pero tenemos la garantía de que se resolverá. Los objetos de promesa vienen con un método llamado ```.then ()```. Nos permite decir: *"Tengo una promesa, cuando se resuelva, entonces esto es lo que quiero que suceda ..."*.

```js
    const promise = new Promise((resolve, reject) => {
    const res = true;
    // Una operacion asíncrona.
    if (res) {
        resolve('Resolved!')
    }
    else {
        reject(Error('Error'))
    }
    })

    promise.then((res) => console.log(res), (err) => alert(err))
```

### Crear un objeto Promise

Para crear un ```Promise``` es necesario usar la palabra ```new```

```js
const executorFn = (resolve, reject) => {
  console.log('The executor function of the promise!');
}

const promise = new Promise(executorFn)
```

El constructor del objeto ```Promise``` toma una función, conocida como la **función ejecutora**, como argumento. Esta función es responsable de resolver o rechazar la promesa.

```js
const executorFn = (resolve, reject) => {
  resolve('Resolved!')
}

const promise = new Promise(executorFn)
```

En el ejemplo dado, la promesa siempre se resuelve incondicionalmente por la función ```resolve```. La función de ```reject``` podría usarse para un rechazo.

### then

```.then()``` es una *higher-order function* toma dos funciones ```callback``` como argumentos; a esta funciones les llamamos ```handlers```. Cuando se cumpla la promesa, se invocará al ```handler```.

- El primer ```handler```, a veces llamado ```onFulfilled```, es de éxito y debe contener la lógica para la resolución de la promesa.
- El segundo ```handler```, a veces llamado ```onRejected```, es de errores, y debe contener la lógica del rechazo de la promesa.

```js
const promise = new Promise((resolve, reject) => {
  setTimeout(() => {
    resolve('Result')
  }, 200)
})

promise.then(res => {console.log(res)}, (err) => {alert(err)})
```

Podemos invocar ```.then ()``` con uno, ambos o ninguno de los ```handler```.  Ya que devuelve una Promesa, incluso si una o ambas funciones del controlador están ausentes. Debido a esto, se pueden encadenar múltiples métodos ```.then ()```. Esto se conoce como composición.

```js

const promise = new Promise(resolve => setTimeout(() => resolve('dAlan'), 100))

promise
.then(res => {
  return res === 'Alan' ? Promise.resolve('Hey Alan!') : Promise.reject('Who are you?')
})
.then(res => {console.log(res)}, err => {alert(err)})

```

### catch

La función pasada como segundo argumento a un método ```.then ()``` de un objeto de ```Promise``` se usa cuando la promesa se rechaza.
Una alternativa a este enfoque es utilizar el método JavaScript ```.catch ()``` del objeto de ```Promise```.

Usamos ```.catch ()``` con una devolución de llamada del ```handler``` de *error* que contiene la lógica de lo que debería suceder si una promesa se rechaza.

```js

const promise = new Promise((resolve, reject) => {  
  setTimeout(() => {
    reject(Error('Promise Rejected Unconditionally.'));
  }, 1000);
});

promise
.then((res) => {
  console.log(value);
})
.catch((err) => {
  alert(err);
})

```

## Fetch

Fetch esta completamente basada en Promise; se usa para acceder y manipular solicitudes y respuestas dentro ```HTTP```, obteniendo recursos de forma asíncrona a través de una red.

Una solicitud ```fetch()``` básica aceptará un parámetro de URL, enviará una solicitud y contendrá un `handler` de éxito y error.

```js

fetch('url')
.then(
  response  => {
    console.log(response);
  },
 rejection => {
    console.error(rejection.message);
);

```

- Primero, llama a la función ```fetch()``` y pásalo una URL como un `string`

```js

fetch('https://api-to-call.com/endpoint');

```

- Encadena un método ```.then ()``` al final de la función ```fetch ()``` y pasa un ```callback``` como primer argumento.

```js

fetch('https://api-to-call.com/endpoint')
.then(response => {});

```

- El método ```.json()``` resolverá una promesa devuelta a un objeto JSON, analizando el texto del cuerpo como JSON. 

```js

fetch('url-that-returns-JSON')
.then(response => response.json())
.then(jsonResponse => {
  console.log(jsonResponse);
});

```

Esto devuelve una promesa que se resuelve en un cuerpo de respuesta con formato JSON como un objeto JavaScript.
